src/qibosoq/programs/base.py

```diff
@@ -159,60 +160,78 @@ def add_pulse_to_register(self, pulse: Pulse):
         else:
             raise NotImplementedError(f"Shape {pulse} not supported!")
 
+    def execute_drive_pulse(self, pulse: Pulse, last_pulse_registered: Dict):
+        """Register a drive pulse if needed, then trigger the respective DAC.
+
+        A pulse gets register if:
+        - it didn't happen in `initialize` (`self.pulses_registered` is False)
+        - it is not identical to the last pulse registered
+
+        """
+        if not self.pulses_registered and (
+            pulse.dac not in last_pulse_registered or pulse != last_pulse_registered[pulse.dac]
+        ):
+            self.add_pulse_to_register(pulse)
+            last_pulse_registered[pulse.dac] = pulse
+        self.pulse(ch=pulse.dac)
+
+    def execute_readout_pulse(
+        self, pulse: Pulse, muxed_pulses_executed: List[Pulse], muxed_ro_executed_indexes: List[int]
+    ):
+        """Register a readout pulse and perform a measurement."""
+        adcs = []
+        if self.is_mux:
+            if pulse not in muxed_pulses_executed:
+                for idx, mux_time in enumerate(self.multi_ro_pulses):
+                    if pulse in mux_time:
+                        idx_mux = idx
+                        break
```

Comment:
This one here you're look for an index, so you could try something like:
```suggestion
                idx_mux = next(idx for idx, mux_time in enumerate(self.multi_ro_pulses) if pulse in mux_time)
```
In the same spirit of the condition above.

Reply:
If you want to have a default, there is a convenient way of doing even that
https://stackoverflow.com/a/9542768/8653979